### PR TITLE
[AIE2][ValueTracking] Add support for AIE2 Intrinsics

### PR DESCRIPTION
**Pull request by andcarminati**
With this change, BasicAA can handle AIE2 intrinsics, applying all existent algorithms (aliasPHI, for example).

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -5939,6 +5956,9 @@ const Value *llvm::getUnderlyingObject(const Value *V, unsigned MaxLookup) {
           V = RP;
           continue;
         }
+      } else if (auto *AIEObject = getUnderlyingObjectAIEIntrinsic(V)) {
```
Comment (gbossu) [inline llvm/lib/Analysis/ValueTracking.cpp:5959]: Could you summarize how that differs from our previous approach of having a separate AA pass for AIE intrinsics?

Comment (andcarminati) [reply llvm/lib/Analysis/ValueTracking.cpp:5959]: Hi @gbossu, basically, our pass can handle just linear code, in this way, we can decompose only in upwards direction. There is no simple way to, for example, switch between different analysis implementation when we don't understand a specific node (PHI, for example). Once we are running the BasicAA, and we reach a PHI node, for example, we cannot easily decide which AA in the chain can handle all the incoming values from that node. 

One approach could be the propagation of the PHI handling inside our AA, however we will definitely handle PHI chains twice (AIEAA an BasicAA)  for the cases containing our specific intrinsics.

Comment (andcarminati) [reply llvm/lib/Analysis/ValueTracking.cpp:5959]: I feel that it is better to have one heavy AA that can handle everything than two heavy AAs. If the first fail to prove NoAlias, we need to run everything again (next AA).

Comment (gbossu) [reply llvm/lib/Analysis/ValueTracking.cpp:5959]: I think I agree. Let's discuss with @konstantinschwarz but I think we can then add a second commit that removes our AIE-specific AA pass.

Comment (gbossu) [reply llvm/lib/Analysis/ValueTracking.cpp:5959]: Let's delay the removal of our custom AA wrapper pass. We might anyway need it for address space disambiguation.

**Review by gbossu — APPROVED** [review]
LGTM :)

**Review by martien-de-jong — APPROVED** [review]
I wonder whether we could get this functionality in a generic hook next to the address space  ones, and include both items in BasicAA upstream.

**Comment by andcarminati** [comment]
> I wonder whether we could get this functionality in a generic hook next to the address space ones, and include both items in BasicAA upstream.

I think we can address this in the future, as a follow-up work.
